aten/src/ATen/native/Sorting.cpp

```diff
@@ -153,34 +209,34 @@ std::tuple<Tensor&, Tensor&> kthvalue_out_cpu(
     indices.zero_();
     return std::forward_as_tuple(values, indices);
   }
-  auto tmp_values = self.clone();
-  auto tmp_indices = at::empty(self.sizes(), self.options().dtype(kLong));
   AT_DISPATCH_ALL_TYPES(self.scalar_type(), "kthvalue_cpu", [&] {
-    dim_apply(
-        {tmp_values, tmp_indices, values, indices},
-        dim,
-        [&](int64_t i, TensorList tl) {
-          auto tmp_values = tl[0].accessor<scalar_t, 1>();
-          auto tmp_indices = tl[1].accessor<int64_t, 1>();
-          scalar_t* mode_value = tl[2].data<scalar_t>();
-          int64_t* mode_index = tl[3].data<int64_t>();
-          for (int64_t j = 0; j < tmp_indices.size(0); j++) {
-            tmp_indices[j] = j;
-          }
-          // we want NaN to be sorted as top for numpy compatibility
-          quick_select_template(
-              tmp_values,
-              k - 1,
-              [](scalar_t x, scalar_t y) -> bool {
-                return ((_isnan<scalar_t>(x) && !_isnan<scalar_t>(y)) || (x > y));
-              },
-              [&](int64_t i, int64_t j) {
-                std::swap(tmp_values[i], tmp_values[j]);
-                std::swap(tmp_indices[i], tmp_indices[j]);
-              });
-          *mode_value = tmp_values[k - 1];
-          *mode_index = tmp_indices[k - 1];
-        });
+    bool can_use_small_index = self.size(dim) <= std::numeric_limits<int_same_size_t<scalar_t>>::max();
+    if (can_use_small_index) {
+      using index_t = int_same_size_t<scalar_t>;
+      qsel_with_indices<scalar_t, index_t>(self, values, indices,
+          [k](const scalar_t *vals, const Tensor& val_slice) {
+            *val_slice.data<scalar_t>() = vals[k - 1];
```

Comment:
`data<>()` is a checked pointer access, which I think is inappropriate here because I believe you're calling this in a loop. However, I don't see an `unsafe_data` member on Tensor, I guess we should add one.

Reply:
It is a little shocking that you're passing an entire honking `Tensor` reference into this function; that sounds like a bunch of unnecessary allocations, especially when you just want to poke some memory at one location.